### DISCUSSION TITLE
mpv串流播放按钮

### DISCUSSION THREAD

**Discussion by Lee-7723** [AlistGo/alist#724] [Show and tell]
分享一个自己写的脚本，将视频下方几个拉取到本地播放器串流播放的按钮中的IINA换成mpv

![image](https://user-images.githubusercontent.com/52600662/157679206-206bf6d6-debb-438e-8241-d6f8c507c601.png)
---------------------------------
***

**1. 添加一个“关联URI协议”**
---------------------------------
参考[Microsoft文档](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)?redirectedfrom=MSDN)

```
Windows Registry Editor Version 5.00

[HKEY_CLASSES_ROOT\mpv-alist]
@="URL:Mpv-alist Protocol"
"URL Protocol"=""

[HKEY_CLASSES_ROOT\mpv-alist\DefaultIcon]
@="F:\\Program Files\\mpv\\mpv.exe,1"

[HKEY_CLASSES_ROOT\mpv-alist\shell]

[HKEY_CLASSES_ROOT\mpv-alist\shell\open]

[HKEY_CLASSES_ROOT\mpv-alist\shell\open\command]
@="\"F:\\Program Files\\mpv\\mpv-alist.bat\" \"%1\""
```
将其中的mpv对应的路径换为你的路径，保存为.reg文件，添加注册表
***

**2. mpv-alist.bat文件内容（mpv已添加至系统环境路径）**
---------------------------------
```batch
set url=%1
mpv "%url:~13%
```
此bat文件路径对应reg注册表文件的最后一行。
***

**3. 在Alist管理中，设置->自定义body中添加如下脚本**
--------------------------------------------------------------
```JavaScript
<script>
function sleep (time) {
    return new Promise((resolve) => setTimeout(resolve, time));
}
var buttons = 0;
async function wait_loading() {
    for (var i = 0; !buttons[5]; i) {
        await sleep(250);
        buttons = document.getElementsByClassName("chakra-button");
    }
}
async function change_button() {
    await wait_loading();
    buttons[5].innerHTML = "mpv";
    buttons[5].href = buttons[5].href.replace("iina://weblink?url=","mpv-alist://");
}
change_button()
</script>
```
浏览器通过mpv-alist://xxxxxxxx来唤起mpv-alist.bat，进一步唤起mpv

**Comment by Lee-7723** [comment]
PC上串流到VLC同理，而且已经提供了vlc的按钮，所以只需要一个bat和一个reg
***
**1. 添加一个“关联URI协议”**
-------------------------------
将下面代码保存为.reg文件，添加注册表（记得修改两处路径）
```
Windows Registry Editor Version 5.00

[HKEY_CLASSES_ROOT\vlc]
"URL Protocol"=""
@="URL:VLC"

[HKEY_CLASSES_ROOT\vlc\DefaultIcon]
@="\"F:\\Program Files\\VLC\\vlc.exe\",0"

[HKEY_CLASSES_ROOT\vlc\shell]

[HKEY_CLASSES_ROOT\vlc\shell\open]

[HKEY_CLASSES_ROOT\vlc\shell\open\command]
@="\"F:\\Program Files\\VLC\\vlc-browser.bat\" \"%1\""
```
***
**2. vlc-browser.bat文件内容（vlc已添加至系统环境路径）**
-------------------------------------------------------------------
注意对应上面注册表文件最后一行的路径
```batch
set url=%1
vlc "%url:~7%
```

**Reply by rui-ao** [reply]
请问如何确定vlc已经添加到系统环境路径了呢，我按照老哥的设置后只是一闪而过一个窗口，vlc并未启动

**Reply by Lee-7723** [reply]
开个命令行窗口，直接输入vlc看看能不能调出来

**Reply by rui-ao** [reply]
> 开个命令行窗口，直接输入vlc看看能不能调出来

CMD窗口中输入vlc以及VLC均能打开软件请问是我哪里没有修改吗

Windows Registry Editor Version 5.00

```
[HKEY_CLASSES_ROOT\vlc]
"URL Protocol"=""
@="URL:VLC"

[HKEY_CLASSES_ROOT\vlc\DefaultIcon]
@="\"C:\\Program Files\\VideoLAN\\VLC\\vlc.exe\",0"

[HKEY_CLASSES_ROOT\vlc\shell]

[HKEY_CLASSES_ROOT\vlc\shell\open]

[HKEY_CLASSES_ROOT\vlc\shell\open\command]
@="\"C:\\Program Files\\VideoLAN\\VLC\\vlc-browser.bat\" \"%1\""
```
然后vlc-browser.bat直接复制老哥的放到了C:\\Program Files\\VideoLAN\\VLC\\vlc-browser.bat位置，请问哪里设置错了吗

**Reply by Lee-7723** [reply]
隔得时间太久了，现在都不太记得了，我这里不太好试了。试试浏览器地址栏里输入vlc:// 能不能调出vlc来。

**Reply by rui-ao** [reply]
> 隔得时间太久了，现在都不太记得了，我这里不太好试了。试试浏览器地址栏里输入vlc:// 能不能调出vlc来。

输入vlc://之后浏览器提示是否允许调用vlc-browser.bat，允许后还是一闪而过，没有启动vlc

**Reply by Lee-7723** [reply]
在bat最后加一行pause
应该会留住黑框，看一下报错是什么

**Reply by rui-ao** [reply]
> 在bat最后加一行pause 应该会留住黑框，看一下报错是什么

![_20221113220106](https://user-images.githubusercontent.com/46027934/201525764-5ddcbff9-fc43-4519-9d64-00b041ed535c.png)
好像还是vlc没识别，但是复制这个命令后我默认打开的cmd窗口可以正常使用，在system这里不能使用，感觉还是我系统变量的问题

**Reply by rui-ao** [reply]
> > 在bat最后加一行pause 应该会留住黑框，看一下报错是什么
> 
> ![_20221113220106](https://user-images.githubusercontent.com/46027934/201525764-5ddcbff9-fc43-4519-9d64-00b041ed535c.png) 好像还是vlc没识别，但是复制这个命令后我默认打开的cmd窗口可以正常使用，在system这里不能使用，感觉还是我系统变量的问题
尝试了许久，自己打开的窗口可以随意使用vlc命令，但是使用注册表打开的就会提示不存在命令，解决办法是bat文件中使用vlc的完整路径比如我的
```
set url=%1
"C:\Program Files\VideoLAN\VLC\vlc.exe" "%url:~7%
```

**Reply by Lee-7723** [reply]
哦确实，我思维定式了，这样直接省略掉环境路径一步

**Reply by shaojs321** [reply]
老哥，我按照你们的方法解析出来的链接，http少个：啊
<img width="463" alt="vlc" src="https://github.com/user-attachments/assets/1667a9bd-6315-4729-bd85-2d1b841211e8" />

**Reply by zitons** [reply]
@shaojs321 https://github.com/AlistGo/alist/issues/7360#issuecomment-2439544206
这个可以试试，不过我没成功，我的方案是bat最后一行换成
`"C:\Program Files\VideoLAN\VLC\vlc.exe" "%url:vlc://http=http:%" `
相当于做了一个替换

**Reply by Alice123qwe** [reply]
@shaojs321
我的方案是直接在bat里硬改成加：的形式，可以试试（我这是https，http把s去掉就行了）

```
@echo off
setlocal enabledelayedexpansion

:: 获取URL
set "input_url=%~1"

:: 转换URL
set "output_url=!input_url:vlc://=!"
set "output_url=!output_url:https//=https://!"

start "" "D:\VLC\vlc.exe" !output_url!
exit

```

**Comment by reddloob** [comment]
更新一下v3.28.0版，0-2是桌面端，XXX换成对应程序的调用方法，3-6是移动端，???换成包名

```
<script>
function sleep (time) {
    return new Promise((resolve) => setTimeout(resolve, time));
}
var buttons = 0;
async function wait_loading() {
    for (var i = 0; !buttons[6]; i) {
        await sleep(250);
        buttons = document.getElementsByClassName("hope-anchor hope-c-iHuheP hope-c-PJLV hope-c-PJLV-idrWMwW-css");
    }
}
async function change_button() {
    await wait_loading();
    buttons[0].href = buttons[0].href.replace("iina://weblink?url=","XXX");
    buttons[1].href = buttons[1].href.replace("potplayer://","XXX");
    buttons[2].href = buttons[2].href.replace("vlc://","XXX");
    buttons[3].href = buttons[3].href.replace(/nplayer-(.+)/, 'intent:$1#Intent;package=???');
    buttons[4].href = buttons[4].href.replace(/infuse:\/\/x-callback-url\/play\?url=(.+)/, 'intent:$1#Intent;package=???');
    buttons[5].href = buttons[5].href.replace("com.mxtech.videoplayer.ad","???");
    buttons[6].href = buttons[6].href.replace("com.mxtech.videoplayer.pro","???");
}
change_button()
</script>
```

**Reply by iamalexblue** [reply]
打扰啦, 请问这个pc端程序的调用方法具体该怎么填?我试过把 "XXX" 替换成可执行程序的路径和自建的 mpv-alist.bat 都不行, 无奈只能上前求助

**Reply by Lee-7723** [reply]
alist的PC端应用上做修改吗？这个不清楚，如果不能注脚本那岂不是要自己重新编译了

**Reply by iamalexblue** [reply]
> alist的PC端应用上做修改吗？这个不清楚，如果不能注脚本那岂不是要自己重新编译了

```
buttons[0].href = buttons[0].href.replace("iina://weblink?url=","XXX");
    buttons[1].href = buttons[1].href.replace("potplayer://","XXX");
    buttons[2].href = buttons[2].href.replace("vlc://","XXX");
    buttons[3].href = buttons[3].href.replace(/nplayer-(.+)/, 'intent:$1#Intent;package=???');
    buttons[4].href = buttons[4].href.replace(/infuse:\/\/x-callback-url\/play\?url=(.+)/, 'intent:$1#Intent;package=???');
    buttons[5].href = buttons[5].href.replace("com.mxtech.videoplayer.ad","???");
    buttons[6].href = buttons[6].href.replace("com.mxtech.videoplayer.pro","???");
```
我猜应该是修改这几行最后的XXX，然后填入 alist管理后台设置内的 `全局header`，但是我不太清楚这个所谓的`程序调用方式`该怎么填☹️

（貌似是回复串行了打扰到了你，抱歉🙏）

**Comment by KnIfER** [comment]
麻烦不，最好能与用户脚本合作。

我电脑常年开启一个局域服务，这个服务可以启动本地应用。

这样，就可以写user.js脚本，直接设置成用外部播放器打开了。

**Comment by kaixinxiaowangzi666** [comment]
在sourceforge里下载mpvnet省事多了

**Reply by KnIfER** [reply]
net ? c#写的壳吗。

我的exe省事，调用mvp的库，图形化缩放移动，全屏等操作。

**Comment by yijunhong666** [comment]
O(∩_∩)O谢谢楼主分享，收下了，真诚感谢！

**Comment by andla119** [comment]
个别版本IINA不能直接替换mpv,url字符转义后mpv读不出来
`set url=%1
mpv %url:mpv-alist://http=http:%`
顺便应付一下浏览器bug

**Comment by zitons** [comment]
分享一下我的mpv-alist.bat文件，这样就不用去alist里面修改了
```@echo off
setlocal enabledelayedexpansion

set "encoded_url=%~1"

for /f "delims=" %%a in ('powershell -command "[System.Uri]::UnescapeDataString('!encoded_url!')"') do (
    set "decoded_url=%%a"
)

set "final_url=!decoded_url!"

echo.
echo Original URL: %encoded_url%
echo Decoded URL: !decoded_url!
echo Final URL: !final_url!
echo.

:: 启动 mpv 播放器
"C:\Users\zitons\mpv\mpv.exe" "!final_url!"

endlocal
pause```顺便说一下mpv的安卓版本真的简陋

**Reply by zitons** [reply]
> 分享一下我的mpv-alist.bat文件，这样就不用去alist里面修改了
> 
> ```
> setlocal enabledelayedexpansion
> 
> set "encoded_url=%~1"
> 
> for /f "delims=" %%a in ('powershell -command "[System.Uri]::UnescapeDataString('!encoded_url!')"') do (
>     set "decoded_url=%%a"
> )
> 
> set "final_url=!decoded_url!"
> 
> echo.
> echo Original URL: %encoded_url%
> echo Decoded URL: !decoded_url!
> echo Final URL: !final_url!
> echo.
> 
> :: 启动 mpv 播放器
> "C:\Users\zitons\mpv\mpv.exe" "!final_url!"
> 
> endlocal
> pause
